### PR TITLE
remake of #357

### PR DESCRIPTION
**Pull request by Adrijaned**
#357 got too messy

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
Bump - there's a hopefully small conflict in this one now, you able to check on it and fix @Adrijaned?

Also, how to even test this one? The clip stacktrace thing should go away? Whereas #388 just would make that issue if still active look less angry, meaning maybe it should be merged first?

File: docs/items.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Item format description:
+Space separated list of individual items ids
+Each id comprised of the following parts (parts surrounded with [] are optional, parts surrounded in <> are to be replaced as per table defined at the end of this document):
+(pay attention to the special chars like | or *)
+[<chance>|][<amount>*]<itemName>[-<slot>][+<itemName>[-<slot>]][+<itemName>[-<slot>]][+<itemName>[-<slot>]]
+The end may optionally continue even with other itemnames.
+chance is a float between 0 and 1, denoting how probable it is ship will receive the item. 0 means not at all probably, 1 means completely sure. Defaults to 1 when not specified
+amount is integer denoting how much of the item player will receive, should the chance check pass. must be >=1, defaults to 1 when not specified
+itemName is the gestalt id of the item, in standard form "moduleName:itemName"
```
Comment (AndyTechGuy) [inline docs/items.txt]: Maybe there should be a temporary note here about the existence of some non-gestalt ids? (The repair item `rep` being the primary example)

File: engine/src/main/resources/assets/schemas/schemaClip.json [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      "required": ["infinite"]
+    },
+    {
+      "properties": {
+        "infinite": {
+          "enum": [
+            false
+          ]
+        },
+        "price": {
```
Comment (AndyTechGuy) [inline engine/src/main/resources/assets/schemas/schemaClip.json]: Should the price have a minimum? I tried putting in a negative price and the game still ran, and I made money by firing bullets then buying more 😄 . Negative price _technically_ works, but I don't think it should be allowed...

File: engine/src/main/resources/assets/schemas/schemaClip.json [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+          "type": "integer",
+          "description": "Buying price of the clip"
+        },
+        "displayName": {
+          "type": "string",
+          "description": "Name of the clip"
```
Comment (AndyTechGuy) [inline engine/src/main/resources/assets/schemas/schemaClip.json]: I included regex patterns for display names in one of my schema fixes (#398), it should probably be added here as well.
```suggestion
          "description": "Name of the clip"
          "pattern": "^[ -~]+$"
```

File: engine/src/main/resources/assets/schemas/schemaClip.json [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        },
+        "plural": {
+          "type": "string",
+          "description": "Plural form of 'displayName'"
```
Comment (AndyTechGuy) [inline engine/src/main/resources/assets/schemas/schemaClip.json]: Same as above
```suggestion
          "description": "Plural form of 'displayName'"
          "pattern": "^[ -~]+$"
```

File: engine/src/main/resources/assets/schemas/schemaClip.json (hunk 2 above)
Comment (AndyTechGuy) [reply engine/src/main/resources/assets/schemas/schemaClip.json]: Also the existence of a price minimum is inconsistant between all schemas, so what is decided should probably be applied to all price fields.

Comment (Adrijaned) [reply engine/src/main/resources/assets/schemas/schemaClip.json]: If you want a module to add enemy-healing bullets that give you money for buying them, why not? :wink: Consistence should however be seeked in the future.

File: docs/items.txt (hunk 1 above)
Comment (Adrijaned) [reply docs/items.txt]: With code modules, better course of action might actually be gestaltifying `rep` and moving it to core.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

File: docs/items.txt (hunk 1 above)
Comment (AndyTechGuy) [reply docs/items.txt]: Okay, in that case after this is merged we should make a separate issue for that. 👍

File: engine/src/main/resources/assets/schemas/schemaClip.json (hunk 2 above)
Comment (AndyTechGuy) [reply engine/src/main/resources/assets/schemas/schemaClip.json]: If negative price should be allowed, can you remove the exlcusive minimums from the other schemas in this PR? I believe they are only in `schemaGun.json` and `schemaAbilityCharges.json`, shouldn't be too much of an issue 😄

File: engine/src/main/java/org/destinationsol/game/ship/ShipBuilder.java [hunk 5]
```diff
@@ -274,8 +274,9 @@ private Hull buildHull(SolGame game, Vector2 position, Vector2 velocity, float a
             doors.add(door);
         }
 
+        Fixture base = getBase(hullConfig.hasBase(), body);
```
Comment (AndyTechGuy) [inline engine/src/main/java/org/destinationsol/game/ship/ShipBuilder.java]: This line is causing the jenkins failure, it must have snuck in during the merge :/

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/resources/assets/schemas/schemaClip.json [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            false
+          ]
+        },
+        "price": {
+          "type": "integer",
+          "description": "Buying price of the clip"
+        },
+        "displayName": {
+          "type": "string",
+          "description": "Name of the clip"
```
Comment (NicholasBatesNZ) [inline engine/src/main/resources/assets/schemas/schemaClip.json]:
```suggestion
          "description": "Name of the clip",
```

File: engine/src/main/resources/assets/schemas/schemaClip.json [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+          "pattern": "^[ -~]+$"
+        },
+        "plural": {
+          "type": "string",
+          "description": "Plural form of 'displayName'"
```
Comment (NicholasBatesNZ) [inline engine/src/main/resources/assets/schemas/schemaClip.json]:
```suggestion
          "description": "Plural form of 'displayName'",
```

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Pull request merged**